File: hw/vizcaino/u3/hw20FarmAnimalSystem/hw20AnimalFarmSystem/src/main/java/ec/edu/espe/animalfarmsystem/model/Pig.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package ec.edu.espe.animalfarmsystem.model;

import ec.edu.espe.animalfarmsystem.controller.IMeatAnimal;
import java.util.ArrayList;
import java.util.Date;

/**
 *
 * @author Adrian Vizcaino, The Softwarriors, @ESPE
 */
public class Pig extends FarmAnimal implements IMeatAnimal {
    private float idealWeight;
    private float premiumCutsPercentage;

    public Pig(float idealWeight, float premiumCutsPercentage, int id, String breed, Date bornOn, float Weight, SlaughterHouse slaughterHouse, Product product, ArrayList<Cut> cuts) {
        super(id, breed, bornOn, Weight, slaughterHouse, product, cuts);
        this.idealWeight = idealWeight;
        this.premiumCutsPercentage = premiumCutsPercentage;
    }

    public boolean isReadyForSlaughter() {
        return this.getWeight() >= this.idealWeight;
    }

    public float calculatePremiumCuts() {
        return (this.getWeight() * this.premiumCutsPercentage) / 100;
    }

    @Override 
    public void sendToSlaughterHouse(SlaughterHouse slaughterHouse) {
        this.slaughterHouse = slaughterHouse;
    }

    @Override
    public void feed(Food food) {
        throw new UnsupportedOperationException("Not supported yet."); 
    }

    @Override
    public ArrayList<Cut> cut() {
        ArrayList<Cut> cutsList = new ArrayList<>();
        float estimatedCutWeight = this.getWeight() / 4;

        cutsList.add(new Cut(1, "Pork Loin", "Back section, lean meat cut", estimatedCutWeight));
        cutsList.add(new Cut(2, "Spare Ribs", "Lower rib cage section", estimatedCutWeight));
        cutsList.add(new Cut(3, "Pork Belly", "Fatty boneless cut from the abdomen", estimatedCutWeight));
        cutsList.add(new Cut(4, "Boston Butt", "Upper shoulder section cut", estimatedCutWeight));
        
        this.setCuts(cutsList);
        return cutsList;
    }

    /**
     * @return the idealWeight
     */
    public float getIdealWeight() {
        return idealWeight;
    }

    /**
     * @param idealWeight the idealWeight to set
     */
    public void setIdealWeight(float idealWeight) {
        this.idealWeight = idealWeight;
    }

    /**
     * @return the premiumCutsPercentage
     */
    public float getPremiumCutsPercentage() {
        return premiumCutsPercentage;
    }

    /**
     * @param premiumCutsPercentage the premiumCutsPercentage to set
     */
    public void setPremiumCutsPercentage(float premiumCutsPercentage) {
        this.premiumCutsPercentage = premiumCutsPercentage;
    }

    @Override
    public String toString() {
        return "Pig{" + super.toString() + ", idealWeight=" + idealWeight + ", premiumCutsPercentage=" + premiumCutsPercentage + '}';
    }
}

File: hw/vizcaino/u3/hw21FASPyJs/hw21Python/model/pig.py
from model.farm_animal import FarmAnimal
from controller.meat_animal import IMeatAnimal
from model.cut import Cut


class Pig(FarmAnimal, IMeatAnimal):
    def __init__(
        self,
        ideal_weight,
        premium_cuts_percentage,
        id,
        breed,
        born_on,
        weight,
        slaughter_house,
        product,
        cuts,
    ):
        super().__init__(
            id,
            breed,
            born_on,
            weight,
            slaughter_house,
            product,
            cuts,
        )

        self.ideal_weight = ideal_weight
        self.premium_cuts_percentage = premium_cuts_percentage

    def is_ready_for_slaughter(self):
        return self.weight >= self.ideal_weight

    def calculate_premium_cuts(self):
        return (
            self.weight * self.premium_cuts_percentage
        ) / 100

    def send_to_slaughter_house(self, slaughter_house):
        self.slaughter_house = slaughter_house

    def feed(self, food):
        raise NotImplementedError()

    def cut(self):
        estimated_cut_weight = self.weight / 4

        self.cuts = [
            Cut(
                1,
                "Pork Loin",
                "Back section, lean meat cut",
                estimated_cut_weight,
            ),
            Cut(
                2,
                "Spare Ribs",
                "Lower rib cage section",
                estimated_cut_weight,
            ),
            Cut(
                3,
                "Pork Belly",
                "Fatty boneless cut from the abdomen",
                estimated_cut_weight,
            ),
            Cut(
                4,
                "Boston Butt",
                "Upper shoulder section cut",
                estimated_cut_weight,
            ),
        ]

        return self.cuts